### PR TITLE
fix: avoid duplicate calls on withdraw hook when sender is also the recipient

### LINKED ISSUE #822: Rename "onLockupStreamWithdrawn()" hooks to avoid duplicate calls
### Discussed in https://github.com/sablier-labs/v2-core/discussions/804

<div type='discussions-op-text'>

<sup>Originally posted by **smol-ninja** January 26, 2024</sup>
A note to the reader: **The following doesn't affect the deployed version of Sablier contracts**.

## Context
The recent [PR: Make withdraw function callable by any account](https://github.com/sablier-labs/v2-core/issues/731) enables anyone to call `withdraw()` on any stream. Each withdrawal attempts to call `onLockupStreamWithdrawn()` on both sender and recipient if they are contracts. As discussed privately on Slack, it raises a concern when the stream has the same sender and recipient. [^1]

## Problem
We use the same function name `onLockupStreamWithdrawn()` in both `ISablierV2Recipient` and `ISablierV2Sender`. The user is expected to implement `onLockupStreamWithdrawn()` with no information regarding which interface he wants to implement it for.  This leads to duplicate calls on this function when `withdraw()` is triggered by a third party and the sender and recipient are the same contract addresses.

[SablierV2Lockup.sol#L278C1-L297C10](https://github.com/sablier-labs/v2-core/blob/staging/src/abstracts/SablierV2Lockup.sol#L278C1-L297C10)
```solidity
if (msg.sender != recipient && recipient.code.length > 0) {
    try ISablierV2Recipient(recipient).onLockupStreamWithdrawn({.. /* snip */ ..}) { } catch { }
}

if (msg.sender != sender && sender.code.length > 0) {
    try ISablierV2Sender(sender).onLockupStreamWithdrawn({.. /* snip */ ..}) { } catch { }
}
```

### Option 1
Rename `onLockupStreamWithdrawn()`
- to `callSenderOnLockupStreamWithdrawn()` in `ISablierV2Sender`
- to `callRecipientOnLockupStreamWithdrawn()` in `ISablierV2Recipient` 

Because we have other functions `OnLockupStreamCanceled()` and `onLockupStreamRenounced()` in the `ISablierV2Recipient`, we should also consider renaming them to follow the same naming convention.

### Option 2
When the sender is the same as the recipient, `withdraw()` calls on `onLockupStreamWithdrawn()` only once.

```solidity
if (msg.sender != recipient && recipient.code.length > 0) {
    try ISablierV2Recipient(recipient).onLockupStreamWithdrawn({.. /* snip */ ..}) { } catch { }
}

if (sender != recipient && msg.sender != sender && sender.code.length > 0) {
    try ISablierV2Sender(sender).onLockupStreamWithdrawn({.. /* snip */ ..}) { } catch { }
}
```

[^1]: A user provides liquidity on AMM through a multisig and then locks the liquidity by setting up a stream of LP tokens to the same multisig. The multisig implements `onLockupStreamWithdrawn()` to withdraw liquidity from the AMM every time a `withdraw()` is triggered.</div>

### PR DESCRIPTION
**Pull request by smol-ninja**
Fixes #822

### CODE REVIEW

File: test/integration/shared/lockup/Lockup.t.sol [hunk 1]
```diff
@@ -44,12 +44,20 @@ abstract contract Lockup_Integration_Shared_Test is Base_Test {
     /// @dev Creates the default stream.
     function createDefaultStream() internal virtual returns (uint256 streamId);
 
+    /// @dev Creates the default stream with the provided sender and recipient.
+    function createDefaultStream(address recipient, address sender) internal virtual returns (uint256 streamId);
```
Comment (andreivladbrg) [inline test/integration/shared/lockup/Lockup.t.sol:48]: let's rename this to `createDefaultStreamWithUsers`

File: test/integration/shared/lockup/Lockup.t.sol [hunk 2]
```diff
@@ -44,12 +44,20 @@ abstract contract Lockup_Integration_Shared_Test is Base_Test {
...
+
     /// @dev Creates the default stream but make it not cancelable.
     function createDefaultStreamNotCancelable() internal virtual returns (uint256 streamId);
 
     /// @dev Creates the default stream with the NFT transfer disabled.
     function createDefaultStreamNotTransferable() internal virtual returns (uint256 streamId);
 
+    /// @dev Creates the default stream with recipient as the sender.
+    function createDefaultStreamToSelf(address sender) internal virtual returns (uint256 streamId) {
```
Comment (andreivladbrg) [inline test/integration/shared/lockup/Lockup.t.sol:57]: let's rename this to `createDefaultStreamToSender`

File: test/integration/concrete/lockup/withdraw-hooks/withdrawHooks.t.sol [hunk 3]
```diff
@@ -0,0 +1,260 @@
...
+
+        // Make the withdrawal.
+        lockup.withdraw({ streamId: streamId, to: address(goodRecipient), amount: withdrawAmount });
+    }
+
+    modifier givenSenderIsRecipient() {
+        _;
+    }
+
+    function test_SenderHook_CallerUnknown() external givenSenderIsRecipient {
```
Comment (andreivladbrg) [inline test/integration/concrete/lockup/withdraw-hooks/withdrawHooks.t.sol:176]: let's change the name

we should keep the function name first

**Review by andreivladbrg — CHANGES_REQUESTED** [review]
I like the idea to divide the tests in multiple files to avoid super complicated tree branches.

File: test/integration/concrete/lockup/withdraw-hooks/withdrawHooks.t.sol (hunk 3 above)
Comment (smol-ninja) [reply test/integration/concrete/lockup/withdraw-hooks/withdrawHooks.t.sol:176]: Any suggestions for the name? I thought about it but couldn't come up with names that begins with `test_WithdrawHooks` and fits the description. There were few options like `test_WithdrawHooks_CallerUnknownGiven_SenderIsRecipient` but that would be a very long name.

**Comment by andreivladbrg** [comment]
@smol-ninja i've pushed a commit to address my feedback, lmk if it does look good

**Comment by smol-ninja** [comment]
Yes. Looks good to me.

**Review by andreivladbrg — APPROVED** [review]

**Review by PaulRBerg — CHANGES_REQUESTED** [review]
Feedback in the comments.

Also, could you confirm that setting `count` to 0 will fail the test if there's at least one call made?

Please answer here:

https://github.com/foundry-rs/foundry/discussions/7187

**Comment by smol-ninja** [comment]
> Also, could you confirm that setting count to 0 will fail the test if there's at least one call made?

It fails. See the log below.

Screenshot 2024-02-20 at 13 36 00

File: test/integration/concrete/lockup/withdraw-hooks/withdrawHooks.tree [hunk 4]
```diff
@@ -0,0 +1,21 @@
+withdrawHooks.t.sol
+├── given recipient is different than the sender
+│   ├── when the caller is unknown
+│   │   ├── it should make one hook call to the sender
+│   │   └── it should make one hook call to the recipient
+│   ├── when the caller is an approved third party
+│   │   ├── it should make one hook call to the sender
+│   │   └── it should make one hook call to the recipient
+│   ├── when the caller is the sender
+│   │   ├── it should make zero hook calls to the sender
```
[the remark this answers is not in the document]
Comment (smol-ninja) [reply test/integration/concrete/lockup/withdraw-hooks/withdrawHooks.tree:10]: Sure.

File: test/integration/concrete/lockup/withdraw-hooks/withdrawHooks.t.sol [hunk 5]
```diff
@@ -0,0 +1,284 @@
...
+
+import { Integration_Test } from "../../../Integration.t.sol";
+import { Withdraw_Integration_Shared_Test } from "../../../shared/lockup/withdraw.t.sol";
+
+abstract contract WithdrawHooks_Integration_Concrete_Test is Integration_Test, Withdraw_Integration_Shared_Test {
+    function setUp() public virtual override(Integration_Test, Withdraw_Integration_Shared_Test) {
+        Withdraw_Integration_Shared_Test.setUp();
+    }
+
+    modifier givenSenderAndRecipientDifferent() {
```
[the remark this answers is not in the document]
Comment (smol-ninja) [reply test/integration/concrete/lockup/withdraw-hooks/withdrawHooks.t.sol:16]: will go with `givenDifferentSenderAndRecipient`

[the remark this answers is not in the document]
Comment (smol-ninja) [reply test/integration/concrete/lockup/withdraw-hooks/withdrawHooks.t.sol:16]: Should I also rename `givenSenderIsRecipient` to `givenSameSenderAndRecipient`?

[the remark this answers is not in the document]
Comment (PaulRBerg) [reply test/integration/concrete/lockup/withdraw-hooks/withdrawHooks.t.sol:16]: Sounds good (it would be a logical mirror).

**Comment by PaulRBerg** [comment]
This PR introduced the `withdrawHooks.t.sol` file to separate out the hook tests from the other tests for the `withdraw` function.

I can no longer see that file on `staging`. When and why did you undo this structure, @smol-ninja, @andreivladbrg?

**Comment by PaulRBerg** [comment]
Also, I can't identify where this boolean check is tested. Do we have any concrete tests written for it? Or are we relying on fuzzing?

SCR-20240329-lwkg

**Comment by PaulRBerg** [comment]
It looks like the hook tests have been removed in this PR:

https://github.com/sablier-labs/v2-core/pull/817

But I still don't understand why.

**Comment by smol-ninja** [comment]
> I can no longer see that file on staging. When and why did you undo this structure.

You are right. It got deleted in [LockupTranched PR](https://github.com/sablier-labs/v2-core/pull/817/files). I was the reviewer so its my mistake that I overlooked it. @andreivladbrg can you please confirm that it was an accident or is there a reason that I cannot recall?

> I can't identify where this boolean check is tested

In [withdrawHooks.tree](https://github.com/sablier-labs/v2-core/blob/9194e814796c01f2f6b4cb56c515fb833224b38e/test/integration/concrete/lockup/withdraw-hooks/withdrawHooks.tree), there is a branch that tests it when sender and recipient are same.

**Comment by PaulRBerg** [comment]
Thanks guys for confirming.

And no worries - error is the default state of affairs.

**Pull request merged**